### DISCUSSION TITLE
[@types/chrome] accept async functions as callbacks?

### DISCUSSION THREAD

**Discussion by karlhorky** [DefinitelyTyped/DefinitelyTyped#70099] [Issues with a @types package]
It would be nice if `@types/chrome` would accept `async` functions to allow for compact async code.

eg:

[`src/background.ts`](https://github.com/karlhorky/archive-webpage-browser-extension/blob/main/src/background.ts) (`karlhorky/archive-webpage-browser-extension`)

```ts
chrome.action.onClicked.addListener(async (tab) => {
  // ...
});
```

This code does work at runtime.

With [`@typescript-eslint/no-misused-promises`](https://typescript-eslint.io/rules/no-misused-promises), there is an error:

```
/home/runner/work/archive-webpage-browser-extension/archive-webpage-browser-extension/src/background.ts
Error:   1:37  error  Promise returned in function argument where a void return was expected  @typescript-eslint/no-misused-promises
```

cc @matcarlson @fregante @BenjaminReich

[a comment by an automated account is not shown]

**Comment by nicolas377** [comment]
I don't see a problem with adding async listeners as long as there's no return value expected. That would be a larger-scale change, however. I'm about 95% sold on this, I probably would be 100% if there was some documentation on using Promises in listeners. If you could find that, I would definitely approve a PR. Definitely work on that PR though if possible, that would be really helpful!

**Comment by BenjaminReich** [comment]
Some documentation which cites this is [here](https://developer.chrome.com/docs/extensions/develop/migrate/api-calls#replace-callbacks).

Though the quote, "**many** extension API methods return promises", seems to indicate maybe some don't(?) and furthermore it might be required to support both. I remember for one of [my PRs](https://github.com/DefinitelyTyped/DefinitelyTyped/pulls?q=is%3Apr+author%3ABenjaminReich+is%3Aclosed) recently I ran into some issues when I tried to do this on the `chrome.fileSystemProvider` API. Let me do a small repro and see if I can jog my memory

**Reply by karlhorky** [reply]
Nice, thanks @BenjaminReich !

Looking forward to seeing the reproduction and finding out what the challenges are with implementing this.

**Reply by nicolas377** [reply]
I don't think that necessarily supports this change specifically though. That section is talking about replacing callbacks entirely with a .then() call (or an await statement), whereas this change is about having async callbacks. Those are two completely different things, and while I'm still not against this change, I don't think that section is evidence for this change.

**Reply by karlhorky** [reply]
It appears that @oliverdunk actually [wrote in the Chromium Extensions Google Group](https://groups.google.com/a/chromium.org/g/chromium-extensions/c/HxXY6uXDkEY/m/vGOY8U2HAQAJ) that returning a promise is unsupported in Chrome extension APIs, at least for the `chrome.runtime.onMessage.addListener` callback:

> The subtle issue in the code you shared is that you're using an async function. So while you think you're returning true, behind the scenes you actually have a function that returns a promise. The extension APIs don't support this.

Apparently there is also a Chromium bug about this from Mar 2021 (Priority P2, Severity S4, "assigned" to `ke...@chromium.org`):

- https://issues.chromium.org/issues/40753031

**Reply by karlhorky** [reply]
...Or maybe I misread that and this thread / lack of compatibility is only related to the `chrome.runtime.onMessage.addListener` API?

This bug seems to be further along (maybe already implemented?):

- https://issues.chromium.org/issues/40633657

cc @chromatim

**Reply by nicolas377** [reply]
Looks like that bug you said was further along is blocked by another migration internal to Chromium, and the bug itself is blocking to the first one you mentioned. I'm retracting my semi-support and turning it into opposition, since this seems to be (albeit quietly) not supported, and while it could work in certain cases, adding support to all events could cause developers using the types to write code that doesn't work as expected due to a bad type.

**Reply by karlhorky** [reply]
Sounds good for now, yep. 

Maybe we'll get lucky in the next weeks/months and one of the Chromium people will chime in with the overall temperature about passing async functions as callbacks (eg. whether it will be supported, when, etc)

**Reply by oliverdunk** [reply]
Hey all! As @nicolas377 pointed out, there are two slightly different features here:

1. Does a call to an extension API return a promise if no callback is passed? [Since February](https://developer.chrome.com/docs/extensions/whats-new#chrome_122_promise_support_on_asynchronous_extension_apis), that has been the case for all methods with a compatible surface. There are just a few exceptions like `desktopCapture.chooseDesktopMedia` since that fires a callback with two parameters, which has no direct equivalent when using promises.
2. If one of the parameters passed to a callback is a function the developer can use to respond, and the developer returns a promise from that callback function instead, what happens?

As mentioned, (1) is implemented today. It also seems to be correctly typed in chrome-types, since each API method has a signature for the callback variant and a signature for the promise one.

Currently, I don't believe we support (2) anywhere. Most API methods don't let the developer respond, so we likely won't ever make changes there and the risk is low. The cases that come to mind as higher risk are runtime.onMessage (we'd like to support promises here, and just need to implement it) and webRequest.onAuthRequired (we're slightly more hesitant here).

Personally, I would lean against the change proposed here (supporting promise returns from functions that don't support it) because of the short-term risk of developer confusion and the longer term risk of it encouraging code that may break if we add promise support to specific methods in the future.

**Reply by nicolas377** [reply]
Looks like we got lucky a lot earlier than expected 😆 On a real note, thanks for weighing in and giving a better perspective on what's going on. I think that's pretty conclusive on this change for now, if anything changes such as the features discussed being added, please feel free to ping me or any of the DT people on this chain, I'd love for this to be a part of extensions, but only if Chromium adds it first.
